### PR TITLE
feat(transformation): adding geo location function support

### PR DESCRIPTION
**Pull request by Jayachand**
## Description of the change

> adding geo location function support

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/util/customTransformer.js [hunk 1]
```diff
@@ -78,6 +79,26 @@ async function runUserTransform(
...
   );
 
+  await jail.set(
+    '_getGeoLocation',
+    new ivm.Reference(async (resolve, ...args) => {
+      try {
+        const geoStartTime = new Date();
+        let geoData = {};
+        if (GEO_RUDDERSTACK_URL && args.length > 0) {
+          const res = await fetch(`${GEO_RUDDERSTACK_URL}/geoip/${args[0]}`);
```
Comment (sivashanmukh) [inline src/util/customTransformer.js:89]: Lets use fetchWithTimeout with timeout set to 99.9p value or the max value of latencies we have seen on geolocation service?

File: src/util/ivmFactory.js [hunk 2]
```diff
@@ -205,6 +206,26 @@ async function createIvm(code, libraryVersionIds, versionId, secrets, testMode)
...
+        if (GEO_RUDDERSTACK_URL && args.length > 0) {
+          const res = await fetch(`${GEO_RUDDERSTACK_URL}/geoip/${args[0]}`);
+          if (res.status === 200) {
+            geoData = await res.json();
+          }
+        }
+        stats.timing('geo_call_duration', geoStartTime, { versionId });
+        resolve.applyIgnored(undefined, [new ivm.ExternalCopy(geoData).copyInto()]);
+      } catch (error) {
+        resolve.applyIgnored(undefined, [new ivm.ExternalCopy({}).copyInto()]);
```
Comment (abhimanyubabbar) [inline src/util/ivmFactory.js:224]:
```suggestion
      let geolocation = {}
      try {
        const start = new Date();
        if (GEO_RUDDERSTACK_URL && args.length > 0) {
          const res = await fetch(`${GEO_RUDDERSTACK_URL}/geoip/${args[0]}`);
          if (res.status === 200) {
            geolocation = await res.json();
          }
        }
        stats.timing('geolocation_call_duration', start, { versionId });
      
      } catch (error) {
        log.debug(...);
      }
      
      resolve.applyIgnored(undefined, [new ivm.ExternalCopy(geoData).copyInto()]);
```

File: src/util/customTransformer.js [hunk 3]
```diff
@@ -78,6 +79,26 @@ async function runUserTransform(
     }),
   );
 
+  await jail.set(
+    '_getGeoLocation',
```
Comment (abhimanyubabbar) [inline src/util/customTransformer.js:83]: Given the implementations are same here and in ivmFactory, should we think about refactoring it into a shared implementation somewhere ?

Comment (abhimanyubabbar) [inline src/util/customTransformer.js:83]: I would name the function `geolocation(ip)` so it's more crisp when customer invokes it like `await geolocation(ip)`. Wdyt ?

Comment (Jayachand) [reply src/util/customTransformer.js:83]: we are deprecating this code.

File: src/util/customTransformer.js (hunk 1 above)
Comment (Jayachand) [reply src/util/customTransformer.js]: done

File: src/util/customTransformer.js [hunk 4]
```diff
@@ -78,6 +79,26 @@ async function runUserTransform(
...
+        if (GEO_RUDDERSTACK_URL && args.length > 0) {
+          const res = await fetch(`${GEO_RUDDERSTACK_URL}/geoip/${args[0]}`);
+          if (res.status === 200) {
+            geoData = await res.json();
+          }
+        }
+        stats.timing('geo_call_duration', geoStartTime, { versionId });
+        resolve.applyIgnored(undefined, [new ivm.ExternalCopy(geoData).copyInto()]);
+      } catch (error) {
+        resolve.applyIgnored(undefined, [new ivm.ExternalCopy({}).copyInto()]);
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply src/util/customTransformer.js]: going with throwing error when failure happens

File: src/util/ivmFactory.js (hunk 2 above)
Comment (Jayachand) [reply src/util/ivmFactory.js]: done

File: src/util/customTransformer.js (hunk 3 above)
Comment (Jayachand) [reply src/util/customTransformer.js]: agreed

**Review by abhimanyubabbar — APPROVED** [review]
LGTM 👍🏼 🎉

**Review by sivashanmukh — APPROVED** [review]

**Pull request merged**